src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala

```diff
@@ -191,7 +191,22 @@ class GoogleExtensions(
       policies <- resources.toList.traverse { resource =>
         accessPolicyDAO.listAccessPolicies(resource.fullyQualifiedId)
       }
-      _ <- onGroupUpdateRecursive(policies.flatten.map(_.id).toList, visitedGroups)
+
+      // ~ Attempt 1 ~
+      // Separate out (1) figuring out if the access policy has any members and (2) collecting the ones that do and returning just the policy ids
+      flatPols: List[(Boolean, FullyQualifiedPolicyId)] <- policies.flatten.map(pol => accessPolicyDAO.listFlattenedPolicyMembers(pol.id).map(x => (x.isEmpty, pol.id))).sequence
+      filteredList1 = flatPols.collect { case (policyHasNoMembers, policyId) if policyHasNoMembers => policyId }
+
+      // ~ Attempt 2 ~
+      // Use filterA to filter out basted on whether the access policy has any members and then map over the policies to get a list of the policy ids
+      filteredList2 <- policies.flatten.filterA { accessPolicy => accessPolicyDAO.listFlattenedPolicyMembers(accessPolicy.id).map(memberSet => memberSet.isEmpty) }.map(x => x.map(_.id))
+
+      // ~ Attempt 3 ~
+      // Use a traverseFilter to filter based on whether the access policy has any members. traverseFilter filters based on Option, not boolean, so in
+      // this case we have to return None if the member set is empty and Some(id of the access policy) if it is not empty.
+      filteredList3 <- policies.flatten.traverseFilter { accessPolicy: AccessPolicy => accessPolicyDAO.listFlattenedPolicyMembers(accessPolicy.id).map(x => if (x.isEmpty) None else Some(accessPolicy.id)) }
```

Comment:
I like 3 better. I think it's because the `if` tells me where the interesting part is happening, whereas it tends to get tucked away in the other two.

Reply:
I agree that #3 is the easiest for me to read.

Reply:
I think even if attempt 3 is not as readable for people who're not familiar with `traverseFilter`, it's should be more performance than the others since we're looping through `policies` in one pass and picking out the ones we're interested in.

I would also change how we get policies to 
```
policies <- resources.toStream.flatTraverse { resource =>
        accessPolicyDAO.listAccessPolicies(resource.fullyQualifiedId)
      }
```
so that we don't need to `flatten` here

Reply:
I am ok with 3.

In all cases using `listFlattenedPolicyMembers` is overkill... I think it is fine to look only at the direct members and check if that is empty. It is probably pretty edgy where a policy groups contains only empty groups.